### PR TITLE
add RNASeq data simulation based on Splat model

### PR DESCRIPTION
**Pull request by EpigeneMax**
Add the ability to simulate RNA-Seq and single-cell RNA-Seq data.
The implemented algorithm is the so-called [Splat model](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1305-0).

Simulated data is useful for tool validation and is often used in publications.
While several omic data simulation packages exist in the R ecosystem, there are few Python alternatives.
This PR aims at closing this gap.

### CODE REVIEW

File: inmoose/sim/splat.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    if group is None:
+        group = np.zeros(nb_samples, dtype=np.int64)
+    group = np.asarray(group)
+    assert len(group) == nb_samples
+
+    nb_batches = len(np.unique(batch))
+    nb_groups = len(np.unique(group))
+
+    # draw gene means
+    original_means = gamma.rvs(alpha, beta, size=nb_genes, random_state=random_state)
```
Comment (guillaumeap) [inline inmoose/sim/splat.py:156]: Not sure here but is beta really the rate of the distribution if set like that?
Looks like rvs would need a 1/beta value for its scale parameter which is the third parameter, what do you think?

File: inmoose/sim/splat.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+    # draw outliers
+    is_outlier = bernoulli.rvs(outlier_prob, size=original_means.shape, random_state=random_state)
+    outlier_factor = lognorm.rvs(outlier_location, outlier_scale, size=original_means.shape, random_state=random_state)
+    gene_means = np.where(is_outlier, outlier_factor * np.median(original_means), original_means)
+
+    # draw library sizes
+    L = lognorm.rvs(libsize_loc, libsize_scale, size=nb_samples, random_state=random_state)
```
Comment (guillaumeap) [inline inmoose/sim/splat.py:164]: Same here, not sure about the parametrisation of scipy
scale seems to be the first argument (even if another argument is named scale), and loc should be at the loc argument
We can discuss that

File: inmoose/sim/splat.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    cell_means = L * gene_means[:,None] / gene_means.sum()
+
+    # biological coefficients of variation (squared)
+    B2 = (phi + 1/np.sqrt(cell_means))**2 * (bcv_df / chi2.rvs(bcv_df, size=nb_genes, random_state=random_state))[:,None]
+
+    # trended cell means
+    trended_cell_means = gamma.rvs(1/B2, cell_means * B2, random_state=random_state)
+
+    # DE fold-changes between cell types
+    de_lf = [get_lognorm_factors(nb_genes, 1, 0.3+0.2*i, 0.1 + i*0.5, 1, random_state=random_state) for i in range(nb_groups)]
```
Comment (guillaumeap) [inline inmoose/sim/splat.py:176]: Previous comment impact also get_lognorm_function

**Review by guillaumeap — COMMENTED** [review]
Cool implementation!
I just added some comments because scipy parametrisation looks different than R, we can discuss it

File: inmoose/sim/splat.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    # cell type specific cell means
+    de_trended_cell_means = de_lf * trended_cell_means
+
+    # batch effect factors
+    be_fact = [get_lognorm_factors(nb_genes, 1, 0.5, 0.1 + 0.5*i, 1, random_state=random_state) for i in range(nb_batches)]
+    be_fact = np.choose(batch, [x[:,None] for x in be_fact])
+    batch_cell_means = be_fact * de_trended_cell_means
+
+    # counts
+    y = poisson.rvs(batch_cell_means, random_state=random_state)
```
Comment (guillaumeap) [inline inmoose/sim/splat.py:188]: We saw that with Kader, is this model creating Zero Inflated Poisson distribution instead on Negative Binomial?

Comment (EpigeneMax) [reply inmoose/sim/splat.py:188]: Yes and no ;-)
The Splat model creates a zero-inflated Poisson distribution, whose parameter follows a Gamma distribution (modified to account for outliers, batches...). Such Poisson-Gamma mixture is in fact a negative binomial distribution (see [the maths here](https://en.wikipedia.org/wiki/Negative_binomial_distribution#Gamma%E2%80%93Poisson_mixture)). I would say it is another view on the same mathematical object.

It just so happens that outliers, batches, mean-variance dependency... are easier to model in the Poisson-Gamma view.

File: inmoose/sim/splat.py (hunk 1 above)
Comment (EpigeneMax) [reply inmoose/sim/splat.py:156]: Nice catch! `rgamma(N, shape=shape, rate=rate)` in R is indeed equivalent to `gamma.rvs(shape, scale=1/rate, size=N)` in Python.

File: inmoose/sim/splat.py (hunk 2 above)
Comment (EpigeneMax) [reply inmoose/sim/splat.py:164]: Got it.
In fact, R `rlnorm` is parameterized with the log-mean and log-sd, so the correct Python counterpart is `lognorm.rvs(s=scale, scale=exp(loc), size=...)`

**Review by guillaumeap — APPROVED** [review]
A new great tool ported in python!

**Pull request merged**